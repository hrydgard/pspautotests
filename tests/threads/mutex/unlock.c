#include "shared.h"

#define UNLOCK_TEST_SIMPLE(title, mutex, count) { \
	int result = sceKernelUnlockMutex(mutex, count); \
	if (result == 0) { \
		schedf("%s: ", title); \
		schedfMutex(mutex); \
	} else { \
		schedf("%s: Failed (%X)\n", title, result); \
	} \
}

#define UNLOCK_TEST(title, attr, initial, count) { \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	UNLOCK_TEST_SIMPLE(title, mutex, count); \
	sceKernelDeleteMutex(mutex); \
}

#define UNLOCK_TEST_THREAD(title, attr, initial, count) { \
	schedf("%s: ", title); \
	schedulingResult = -1; \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	sceKernelStartThread(lockThread, sizeof(int), &mutex); \
	sceKernelDelayThread(500); \
	int result = sceKernelUnlockMutex(mutex, count); \
	schedf("L2 "); \
	sceKernelDelayThread(500); \
	sceKernelDeleteMutex(mutex); \
	sceKernelWaitThreadEnd(lockThread, NULL); \
	if (result == 0) { \
		schedf("OK (thread=%08X)\n", schedulingResult); \
	} else { \
		schedf("Failed (thread=%08X, main=%08X)\n", schedulingResult, result); \
	} \
	sceKernelTerminateThread(lockThread); \
}

static int lockFunc(SceSize argSize, void* argPointer) {
	SceUInt timeout = 1000;
	schedulingResult = sceKernelLockMutex(*(int*) argPointer, 1, &timeout);
	schedf("L1 ");
	sceKernelDelayThread(1000);
	return 0;
}

static int unlockFunc(SceSize argSize, void* argPointer) {
	int result = sceKernelUnlockMutex(*(int*) argPointer, 1);
	schedf("After unlock: %08X\n", result);
	return 0;
}

int main(int argc, char **argv) {
	UNLOCK_TEST("Unlock 0 => 0", PSP_MUTEX_ATTR_FIFO, 0, 0);
	UNLOCK_TEST("Unlock 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1);
	UNLOCK_TEST("Unlock 0 => 2", PSP_MUTEX_ATTR_FIFO, 0, 2);
	UNLOCK_TEST("Unlock 0 => -1", PSP_MUTEX_ATTR_FIFO, 0, -1);
	UNLOCK_TEST("Unlock 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1);
	UNLOCK_TEST("Unlock 1 => 2", PSP_MUTEX_ATTR_FIFO, 1, 2);
	UNLOCK_TEST("Unlock 2 => 1", PSP_MUTEX_ATTR_FIFO, 2, 1);
	UNLOCK_TEST("Unlock 0 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 2);
	UNLOCK_TEST("Unlock 0 => -1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, -1);
	UNLOCK_TEST("Unlock 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1);
	UNLOCK_TEST("Unlock 1 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 2);

	SceUID lockThread = CREATE_SIMPLE_THREAD(lockFunc);
	UNLOCK_TEST_THREAD("Locked 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1);
	UNLOCK_TEST_THREAD("Locked 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1);
	UNLOCK_TEST_THREAD("Locked 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1);
	UNLOCK_TEST_THREAD("Locked 2 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 2, 1);
	UNLOCK_TEST_THREAD("Locked 1 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 2);
	UNLOCK_TEST_THREAD("Locked 2 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 2, 2);
	UNLOCK_TEST_THREAD("Locked 0 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 1);

	SceUID unlockThread = CREATE_SIMPLE_THREAD(unlockFunc);
	SceUID mutex = sceKernelCreateMutex("unlock", 0, 1, NULL);
	sceKernelStartThread(unlockThread, sizeof(int), &mutex);
	sceKernelDelayThread(500);
	sceKernelDeleteMutex(mutex);

	UNLOCK_TEST_SIMPLE("NULL => 0", 0, 0);
	UNLOCK_TEST_SIMPLE("NULL => 1", 0, 1);
	UNLOCK_TEST_SIMPLE("Invalid => 1", 0xDEADBEEF, 1);
	UNLOCK_TEST_SIMPLE("Deleted => 1", mutex, 1);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelUnlockMutex(0, 1);
	);
	BASIC_SCHED_TEST("Zero",
		result = sceKernelUnlockMutex(mutex2, 0);
	);
	BASIC_SCHED_TEST("Unlock same",
		result = sceKernelUnlockMutex(mutex1, 1);
	);
	BASIC_SCHED_TEST("Unlock other",
		result = sceKernelUnlockMutex(mutex2, 1);
	);

	return 0;
}