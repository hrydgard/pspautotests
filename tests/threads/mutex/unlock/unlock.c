#include "../sub_shared.h"

SETUP_SCHED_TEST;

#define UNLOCK_TEST_SIMPLE(title, mutex, count) { \
	int result = sceKernelUnlockMutex(mutex, count); \
	if (result == 0) { \
		printf("%s: OK\n", title); \
	} else { \
		printf("%s: Failed (%X)\n", title, result); \
	} \
}

#define UNLOCK_TEST(title, attr, initial, count) { \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	UNLOCK_TEST_SIMPLE(title, mutex, count); \
	sceKernelDeleteMutex(mutex); \
}

#define UNLOCK_TEST_THREAD(title, attr, initial, count) { \
	printf("%s: ", title); \
	SceUID mutex = sceKernelCreateMutex("lock", attr, initial, NULL); \
	sceKernelStartThread(lockThread, sizeof(int), &mutex); \
	sceKernelDelayThread(500); \
	int result = sceKernelUnlockMutex(mutex, count); \
	printf("L2 "); \
	sceKernelDelayThread(500); \
	if (result == 0) { \
		printf("OK\n"); \
	} else { \
		printf("Failed (%X)\n", result); \
	} \
	sceKernelDeleteMutex(mutex); \
	sceKernelTerminateThread(lockThread); \
}

static int lockFunc(SceSize argSize, void* argPointer) {
	SceUInt timeout = 1000;
	sceKernelLockMutex(*(int*) argPointer, 1, &timeout);
	printf("L1 ");
	sceKernelDelayThread(1000);
	return 0;
}

static int deleteMeFunc(SceSize argSize, void* argPointer) {
	sceKernelLockMutex(*(int*) argPointer, 1, NULL);
	sceKernelDelayThread(1000);
	int result = sceKernelUnlockMutex(*(int*) argPointer, 1);
	printf("After delete: %08X\n", result);
	return 0;
}

int main(int argc, char **argv) {
	UNLOCK_TEST("Unlock 0 => 0", PSP_MUTEX_ATTR_FIFO, 0, 0);
	UNLOCK_TEST("Unlock 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1);
	UNLOCK_TEST("Unlock 0 => 2", PSP_MUTEX_ATTR_FIFO, 0, 2);
	UNLOCK_TEST("Unlock 0 => -1", PSP_MUTEX_ATTR_FIFO, 0, -1);
	UNLOCK_TEST("Unlock 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1);
	UNLOCK_TEST("Unlock 1 => 2", PSP_MUTEX_ATTR_FIFO, 1, 2);
	UNLOCK_TEST("Unlock 2 => 1", PSP_MUTEX_ATTR_FIFO, 2, 1);
	UNLOCK_TEST("Unlock 0 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 2);
	UNLOCK_TEST("Unlock 0 => -1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, -1);
	UNLOCK_TEST("Unlock 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1);
	UNLOCK_TEST("Unlock 1 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 2);

	SceUID lockThread = CREATE_SIMPLE_THREAD(lockFunc);
	UNLOCK_TEST_THREAD("Locked 1 => 1", PSP_MUTEX_ATTR_FIFO, 1, 1);
	UNLOCK_TEST_THREAD("Locked 0 => 1", PSP_MUTEX_ATTR_FIFO, 0, 1);
	UNLOCK_TEST_THREAD("Locked 1 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 1);
	UNLOCK_TEST_THREAD("Locked 2 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 2, 1);
	UNLOCK_TEST_THREAD("Locked 1 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 1, 2);
	UNLOCK_TEST_THREAD("Locked 2 => 2 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 2, 2);
	UNLOCK_TEST_THREAD("Locked 0 => 1 (recursive)", PSP_MUTEX_ATTR_ALLOW_RECURSIVE, 0, 1);

	SceUID deleteThread = CREATE_SIMPLE_THREAD(deleteMeFunc);
	SceUID mutex = sceKernelCreateMutex("unlock", 0, 0, NULL);
	sceKernelStartThread(deleteThread, sizeof(int), &mutex);
	sceKernelDelayThread(500);
	sceKernelDeleteMutex(mutex);

	UNLOCK_TEST_SIMPLE("NULL => 0", 0, 0);
	UNLOCK_TEST_SIMPLE("NULL => 1", 0, 1);
	UNLOCK_TEST_SIMPLE("Invalid => 1", 0xDEADBEEF, 1);
	UNLOCK_TEST_SIMPLE("Deleted => 1", mutex, 1);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelUnlockMutex(0, 1);
	);
	BASIC_SCHED_TEST("Zero",
		result = sceKernelUnlockMutex(mutex2, 0);
	);
	BASIC_SCHED_TEST("Unlock same",
		result = sceKernelUnlockMutex(mutex1, 1);
	);
	BASIC_SCHED_TEST("Unlock other",
		result = sceKernelUnlockMutex(mutex2, 1);
	);

	return 0;
}