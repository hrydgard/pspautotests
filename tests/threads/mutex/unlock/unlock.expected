Unlock 0 => 0: Failed (800201BD)
Unlock 0 => 1: Failed (800201C5)
Unlock 0 => 2: Failed (800201BD)
Unlock 0 => -1: Failed (800201BD)
Unlock 1 => 1: OK
Unlock 1 => 2: Failed (800201BD)
Unlock 2 => 1: Failed (800201C3)
Unlock 0 => 2 (recursive): Failed (800201C5)
Unlock 0 => -1 (recursive): Failed (800201BD)
Unlock 1 => 1 (recursive): OK
Unlock 1 => 2 (recursive): Failed (800201C7)
Locked 1 => 1: L1 L2 OK
Locked 0 => 1: L1 L2 Failed (800201C5)
Locked 1 => 1 (recursive): L1 L2 OK
Locked 2 => 1 (recursive): L2 L1 OK
Locked 1 => 2 (recursive): L2 L1 Failed (800201C7)
Locked 2 => 2 (recursive): L1 L2 OK
Locked 0 => 1 (recursive): L1 L2 Failed (800201C5)
After delete: 800201C3
NULL => 0: Failed (800201C3)
NULL => 1: Failed (800201C3)
Invalid => 1: Failed (800201C3)
Deleted => 1: Failed (800201C3)
NULL: A1B1C2E2D2F2 (thread=800201B5, main=800201C3)
Zero: A1B1C2E2D2F2 (thread=800201B5, main=800201BD)
Unlock same: A1B1C2D2E2F1 (thread=00000000, main=00000000)
Unlock other: A1B1C2E2D2F2 (thread=800201B5, main=800201C5)
